File: src/turn/LLMStreamReader.ts
/**
 * LLMStreamReader — consumes a single LLM round-trip.
 *
 * Extracted from `Turn.streamOneIteration` (R3 refactor, 2026-04-19).
 * Owns stream event assembly: text_delta / thinking_delta +
 * thinking_signature (T4-18) / tool_use_start + tool_use_input_delta /
 * block_stop / message_end / error. Returns the ordered content blocks
 * ready to be re-emitted to the model on the next round.
 */

import type {
  LLMClient,
  LLMContentBlock,
  LLMEvent,
  LLMMessage,
  LLMStreamRequest,
  LLMToolDef,
  LLMUsage,
} from "../transport/LLMClient.js";
import type { SseWriter } from "../transport/SseWriter.js";
import { RepetitionDetector } from "./RepetitionDetector.js";
import type { SystemBlock } from "../prompt/CacheOptimizedPrompt.js";

export type StopReasonRaw =
  | "end_turn"
  | "tool_use"
  | "max_tokens"
  | "stop_sequence"
  | "refusal"
  | "pause_turn"
  | null;

export interface LLMStreamReaderDeps {
  readonly llm: LLMClient;
  readonly model: string;
  readonly sse: SseWriter;
  readonly abortSignal?: AbortSignal;
  /**
   * Called when the stream emits an `error` event or the upstream
   * connect throws. Receives `(code, err)`; the reader then re-throws
   * the error.
   */
  readonly onError: (code: string, err: unknown) => void;
}

export interface LLMStreamReaderResult {
  blocks: LLMContentBlock[];
  stopReason: StopReasonRaw;
  usage: LLMUsage;
}

/**
 * Run one `/v1/messages` round-trip and materialise the assistant
 * blocks. The returned blocks preserve streamed order; thinking blocks
 * carry their signatures so Anthropic accepts a replay on the next
 * turn (T4-18).
 */
export interface ReadOneOptions {
  /** Override thinking mode for this call (e.g. disable for empty-response recovery). */
  thinkingOverride?: { type: "adaptive" } | { type: "disabled" };
  /** Cross-service diagnostic trace ID propagated to api-proxy. */
  traceId?: string;
  /** Non-authoritative routing metadata passed to LLM transport/proxy logs. */
  routing?: LLMStreamRequest["routing"];
}

const MAX_DOCUMENT_DRAFT_PREVIEW_CHARS = 4_000;

interface ToolInputAccumulator {
  id: string;
  name: string;
  inputJson: string;
}

interface DocumentDraftState {
  contentPreview: string;
  contentLength: number;
  truncated: boolean;
  filename?: string;
}

interface DocumentDraftCandidate {
  filename?: string;
  format: "md" | "txt";
  content: string;
}

function documentFormatFromPath(filePath: string | undefined): "md" | "txt" | null {
  const normalized = (filePath ?? "").split(/[?#]/, 1)[0]?.toLowerCase() ?? "";
  if (normalized.endsWith(".md") || normalized.endsWith(".markdown")) return "md";
  if (normalized.endsWith(".txt")) return "txt";
  return null;
}

function decodeJsonStringPrefix(raw: string, start: number): string {
  let value = "";
  let escaped = false;
  for (let index = start; index < raw.length; index += 1) {
    const char = raw[index];
    if (escaped) {
      switch (char) {
        case "n":
          value += "\n";
          break;
        case "r":
          value += "\r";
          break;
        case "t":
          value += "\t";
          break;
        case "b":
          value += "\b";
          break;
        case "f":
          value += "\f";
          break;
        case "u": {
          const hex = raw.slice(index + 1, index + 5);
          if (/^[0-9a-f]{4}$/i.test(hex)) {
            value += String.fromCharCode(Number.parseInt(hex, 16));
            index += 4;
          }
          break;
        }
        default:
          value += char ?? "";
          break;
      }
      escaped = false;
      continue;
    }
    if (char === "\\") {
      escaped = true;
      continue;
    }
    if (char === "\"") break;
    value += char ?? "";
  }
  return value;
}

function jsonStringFieldPrefix(raw: string, key: string): string | undefined {
  const escapedKey = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  const pattern = new RegExp(`"${escapedKey}"\\s*:\\s*"`, "g");
  const match = pattern.exec(raw);
  if (!match) return undefined;
  return decodeJsonStringPrefix(raw, match.index + match[0].length);
}

function firstJsonStringFieldPrefix(raw: string, keys: readonly string[]): string | undefined {
  for (const key of keys) {
    const value = jsonStringFieldPrefix(raw, key);
    if (value !== undefined) return value;
  }
  return undefined;
}

function parsedObject(raw: string): Record<string, unknown> | null {
  try {
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" && !Array.isArray(parsed)
      ? parsed as Record<string, unknown>
      : null;
  } catch {
    return null;
  }
}

function stringField(object: Record<string, unknown> | null, keys: readonly string[]): string | undefined {
  if (!object) return undefined;
  for (const key of keys) {
    const value = object[key];
    if (typeof value === "string") return value;
  }
  return undefined;
}

function documentWriteSourceContent(source: unknown): string | undefined {
  if (typeof source === "string") return source;
  if (!source || typeof source !== "object" || Array.isArray(source)) return undefined;
  return stringField(source as Record<string, unknown>, ["content", "markdown", "text"]);
}

function draftCandidateFromTool(tool: ToolInputAccumulator): DocumentDraftCandidate | null {
  if (tool.name !== "FileWrite" && tool.name !== "DocumentWrite") return null;
  const parsed = parsedObject(tool.inputJson);

  if (tool.name === "FileWrite") {
    const filename =
      stringField(parsed, ["path", "file_path", "filepath", "filename"]) ??
      firstJsonStringFieldPrefix(tool.inputJson, ["path", "file_path", "filepath", "filename"]);
    const format = documentFormatFromPath(filename);
    if (!format) return null;
    const content =
      stringField(parsed, ["content"]) ??
      jsonStringFieldPrefix(tool.inputJson, "content") ??
      "";
    return content ? { filename, format, content } : null;
  }

  const filename =
    stringField(parsed, ["filename", "path"]) ??
    firstJsonStringFieldPrefix(tool.inputJson, ["filename", "path"]);
  const formatValue =
    stringField(parsed, ["format"]) ??
    jsonStringFieldPrefix(tool.inputJson, "format");
  const format =
    formatValue === "md" || formatValue === "txt"
      ? formatValue
      : documentFormatFromPath(filename);
  if (format !== "md" && format !== "txt") return null;
  const content =
    documentWriteSourceContent(parsed?.source) ??
    firstJsonStringFieldPrefix(tool.inputJson, ["content", "markdown", "text", "source"]) ??
    "";
  return content ? { filename, format, content } : null;
}

function previewTail(content: string): { contentPreview: string; truncated: boolean } {
  if (content.length <= MAX_DOCUMENT_DRAFT_PREVIEW_CHARS) {
    return { contentPreview: content, truncated: false };
  }
  return {
    contentPreview: content.slice(-MAX_DOCUMENT_DRAFT_PREVIEW_CHARS),
    truncated: true,
  };
}

export async function readOne(
  deps: LLMStreamReaderDeps,
  systemPrompt: string | SystemBlock[],
  messages: LLMMessage[],
  toolDefs: LLMToolDef[],
  options?: ReadOneOptions,
): Promise<LLMStreamReaderResult> {
  throwIfAborted(deps.abortSignal);
  // Per-block accumulators — indexed by blockIndex from the SSE.
  const textByIndex = new Map<number, string>();
  // T4-18: thinking blocks must be preserved across iterations with
  // their signatures so Anthropic accepts the replayed assistant
  // trajectory on subsequent turns. Accumulate text + signature per
  // block, assembled alongside text/tool_use blocks below.
  const thinkingByIndex = new Map<
    number,
    { thinking: string; signature: string }
  >();
  const toolByIndex = new Map<number, ToolInputAccumulator>();
  const documentDraftByIndex = new Map<number, DocumentDraftState>();
  const blockOrder: number[] = [];

  let stream: AsyncGenerator<LLMEvent, void, void>;
  try {
    const systemValue: LLMStreamRequest["system"] = Array.isArray(systemPrompt)
      ? systemPrompt
      : systemPrompt || undefined;
    stream = deps.llm.stream({
      model: deps.model,
      system: systemValue,
      messages,
      tools: toolDefs.length ? toolDefs : undefined,
      ...(deps.abortSignal ? { signal: deps.abortSignal } : {}),
      ...(options?.thinkingOverride ? { thinking: options.thinkingOverride } : {}),
      ...(options?.routing ? { routing: options.routing } : {}),
    });
  } catch (err) {
    deps.onError("llm_connect_failed", err);
    throw err;
  }

  let stopReason: StopReasonRaw = null;
  let usage: LLMUsage = { inputTokens: 0, outputTokens: 0 };
  let repetitionAborted = false;

  // Streaming repetition detector — catches LLM text degeneration
  // (same sentence repeating indefinitely within a single response).
  const repetitionDetector = new RepetitionDetector();

  for await (const evt of stream) {
    throwIfAborted(deps.abortSignal);
    switch (evt.kind) {
      case "text_delta": {
        const prev = textByIndex.get(evt.blockIndex) ?? "";
        if (prev.length === 0) blockOrder.push(evt.blockIndex);
        textByIndex.set(evt.blockIndex, prev + evt.delta);

        // Check for degenerate repetition before emitting to client.
        const repResult = repetitionDetector.feed(evt.delta);
        if (repResult.detected) {
          console.warn(
            `[core-agent] REPETITION DETECTED — aborting stream.` +
            ` pattern="${repResult.pattern}" count=${repResult.count}`,
          );
          deps.sse.agent({
            type: "text_delta",
            delta: "\n\n⚠️ [반복 감지됨 — 응답이 중단되었습니다]",
          });
          repetitionAborted = true;
          break;
        }

        // Emit on the structured `event: agent` channel only. The legacy
        // OpenAI-compat `choices[0].delta.content` path was previously
        // dual-emitted here as a migration aid for pre-§7.9 clients; all
        // live clients (web/mobile chat-client, Telegram/Discord capture
        // writer) now consume the agent channel, and dual-emit caused
        // every token to render twice once the client wired text_delta
        // (commit eda9047c, 2026-04-20). `legacyDelta` stays on the ABI
        // for `legacyFinish` (`data: [DONE]`) and any future adapter.
        deps.sse.agent({ type: "text_delta", delta: evt.delta });
        break;
      }
      case "thinking_delta": {
        const prev = thinkingByIndex.get(evt.blockIndex);
        if (!prev) {
          blockOrder.push(evt.blockIndex);
          thinkingByIndex.set(evt.blockIndex, {
            thinking: evt.delta,
            signature: "",
          });
        } else {
          prev.thinking += evt.delta;
        }
        deps.sse.agent({ type: "thinking_delta", delta: evt.delta });
        break;
      }
      case "thinking_signature": {
        // Signature arrives at the end of the thinking block. If no
        // delta preceded it (edge case), still register the block.
        const prev = thinkingByIndex.get(evt.blockIndex);
        if (prev) {
          prev.signature = evt.signature;
        } else {
          blockOrder.push(evt.blockIndex);
          thinkingByIndex.set(evt.blockIndex, {
            thinking: "",
            signature: evt.signature,
          });
        }
        break;
      }
      case "tool_use_start": {
        toolByIndex.set(evt.blockIndex, {
          id: evt.id,
          name: evt.name,
          inputJson: "",
        });
        blockOrder.push(evt.blockIndex);
        // Note: tool_start is NOT emitted here — deferred to runTools
        // so we can include the full input_preview (input is streamed
        // via tool_use_input_delta and not complete until block_stop).
        break;
      }
      case "tool_use_input_delta": {
        const cur = toolByIndex.get(evt.blockIndex);
        if (cur) {
          cur.inputJson += evt.partial;
          const draft = draftCandidateFromTool(cur);
          if (draft) {
            const preview = previewTail(draft.content);
            const previous = documentDraftByIndex.get(evt.blockIndex);
            if (
              !previous ||
              previous.contentPreview !== preview.contentPreview ||
              previous.contentLength !== draft.content.length ||
              previous.filename !== draft.filename
            ) {
              deps.sse.agent({
                type: "document_draft",
                id: cur.id,
                ...(draft.filename ? { filename: draft.filename } : {}),
                format: draft.format,
                contentPreview: preview.contentPreview,
                contentLength: draft.content.length,
                truncated: preview.truncated,
              });
              documentDraftByIndex.set(evt.blockIndex, {
                ...preview,
                contentLength: draft.content.length,
                ...(draft.filename ? { filename: draft.filename } : {}),
              });
            }
          }
        }
        break;
      }
      case "block_stop":
        // Nothing to do; finalization happens when we assemble blocks.
        break;
      case "message_end":
        stopReason = evt.stopReason;
        usage = evt.usage;
        break;
      case "error":
        deps.onError(evt.code, new Error(evt.message));
        throw new Error(`${evt.code}: ${evt.message}`);
    }
    // Break the stream consumption loop if repetition was detected.
    if (repetitionAborted) break;
    throwIfAborted(deps.abortSignal);
  }

  // If repetition was detected, force end_turn so the commit pipeline
  // finalizes cleanly rather than attempting tool dispatch or recovery.
  if (repetitionAborted) {
    stopReason = "end_turn";
  }

  // Assemble blocks in streamed order, de-duped.
  const seen = new Set<number>();
  const blocks: LLMContentBlock[] = [];
  for (const idx of blockOrder) {
    if (seen.has(idx)) continue;
    seen.add(idx);
    const txt = textByIndex.get(idx);
    if (txt !== undefined) {
      blocks.push({ type: "text", text: txt });
      continue;
    }
    // T4-18: assemble thinking block with signature for replay.
    const thinking = thinkingByIndex.get(idx);
    if (thinking !== undefined) {
      blocks.push({
        type: "thinking",
        thinking: thinking.thinking,
        signature: thinking.signature,
      });
      continue;
    }
    const tu = toolByIndex.get(idx);
    if (tu) {
      let input: unknown = {};
      if (tu.inputJson.length > 0) {
        try {
          input = JSON.parse(tu.inputJson);
        } catch {
          input = { _malformed: true, _raw: tu.inputJson };
        }
      }
      blocks.push({ type: "tool_use", id: tu.id, name: tu.name, input });
    }
  }

  return { blocks, stopReason, usage };
}

function throwIfAborted(signal?: AbortSignal): void {
  if (!signal?.aborted) return;
  const reason = (signal as AbortSignal & { reason?: unknown }).reason;
  throw reason instanceof Error ? reason : new Error("llm_stream_aborted");
}

File: openmagi_core_agent/runtime/llm_stream_reader.py
from __future__ import annotations

from collections.abc import AsyncIterable, Callable, Mapping, Sequence
from dataclasses import dataclass, field
import json
import re
from typing import Any, Literal, TypeAlias


StopReason: TypeAlias = Literal[
    "end_turn",
    "tool_use",
    "max_tokens",
    "stop_sequence",
    "refusal",
    "pause_turn",
]
OnError: TypeAlias = Callable[[str, BaseException], None]

_DEFAULT_USAGE = {"inputTokens": 0, "outputTokens": 0}
_ESCALATION_KEYS = frozenset(
    (
        "model",
        "modelLabel",
        "model_label",
        "provider",
        "providerLabel",
        "provider_label",
        "credentialRef",
        "credential_ref",
        "apiKey",
        "api_key",
    ),
)
_SENTENCE_SPLIT_RE = re.compile(r"(?<=[.!?。！？\n])\s*")


@dataclass(frozen=True)
class LLMStreamReaderConfig:
    enabled: bool = False
    local_fake_provider_enabled: bool = False
    document_preview_chars: int = 4_000
    legacy_rendering_enabled: bool = False
    repetition_min_pattern_chars: int = 40
    repetition_threshold: int = 3
    repetition_check_interval_chars: int = 200


@dataclass(frozen=True)
class LLMStreamReaderResult:
    blocks: list[dict[str, Any]] = field(default_factory=list)
    public_events: list[dict[str, Any]] = field(default_factory=list)
    stop_reason: StopReason | None = None
    usage: dict[str, Any] = field(default_factory=lambda: dict(_DEFAULT_USAGE))
    provider_request: dict[str, Any] | None = None
    skipped_reason: str | None = None


class LLMStreamReaderError(RuntimeError):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code
        self.message = message


class LLMStreamAbortError(RuntimeError):
    pass


@dataclass
class _ToolInput:
    id: str
    name: str
    input_json: str = ""


@dataclass(frozen=True)
class _DocumentDraftCandidate:
    format: Literal["md", "txt"]
    content: str
    filename: str | None = None


@dataclass(frozen=True)
class _DocumentDraftState:
    content_preview: str
    content_length: int
    truncated: bool
    filename: str | None = None


@dataclass(frozen=True)
class _RepetitionResult:
    detected: bool
    pattern: str | None = None
    count: int | None = None


class LLMStreamReader:
    def __init__(
        self,
        fake_provider: Any,
        *,
        config: LLMStreamReaderConfig | None = None,
    ) -> None:
        self._provider = fake_provider
        self._config = config or LLMStreamReaderConfig()

    async def read_one(
        self,
        *,
        model: str,
        messages: Sequence[Mapping[str, Any]],
        system_prompt: str = "",
        tool_defs: Sequence[Mapping[str, Any]] | None = None,
        thinking_override: Mapping[str, Any] | None = None,
        trace_id: str | None = None,
        authoritative_model: bool = False,
        routing: Mapping[str, Any] | None = None,
        request_controlled_metadata: Mapping[str, Any] | None = None,
        abort_flag: object | None = None,
        on_error: OnError | None = None,
    ) -> LLMStreamReaderResult:
        if not self._config.enabled:
            return LLMStreamReaderResult(skipped_reason="disabled")
        if not self._config.local_fake_provider_enabled:
            return LLMStreamReaderResult(skipped_reason="local_fake_stream_provider_disabled")
        if getattr(self._provider, "openmagi_local_fake_provider", False) is not True:
            return LLMStreamReaderResult(skipped_reason="local_fake_stream_provider_untrusted")

        _raise_if_aborted(abort_flag)
        request = _build_provider_request(
            model=model,
            messages=messages,
            system_prompt=system_prompt,
            tool_defs=tool_defs or (),
            thinking_override=thinking_override,
            trace_id=trace_id,
            authoritative_model=authoritative_model,
            routing=routing,
            request_controlled_metadata=request_controlled_metadata,
        )

        try:
            stream = self._provider.stream(request)
        except BaseException as exc:
            _call_on_error(on_error, "llm_connect_failed", exc)
            raise LLMStreamReaderError(
                "llm_connect_failed",
                _sanitize_public_text(str(exc)) or "[redacted-provider-error]",
            ) from exc

        return await self._reduce_stream(
            stream,
            request=request,
            abort_flag=abort_flag,
            on_error=on_error,
        )

    async def _reduce_stream(
        self,
        stream: AsyncIterable[Mapping[str, Any]],
        *,
        request: dict[str, Any],
        abort_flag: object | None,
        on_error: OnError | None,
    ) -> LLMStreamReaderResult:
        text_by_index: dict[int, str] = {}
        thinking_by_index: dict[int, dict[str, str]] = {}
        tool_by_index: dict[int, _ToolInput] = {}
        document_draft_by_index: dict[int, _DocumentDraftState] = {}
        block_order: list[int] = []
        public_events: list[dict[str, Any]] = []
        stop_reason: StopReason | None = None
        usage: dict[str, Any] = dict(_DEFAULT_USAGE)
        repetition_aborted = False
        repetition_detector = _RepetitionDetector(
            min_pattern_chars=self._config.repetition_min_pattern_chars,
            threshold=self._config.repetition_threshold,
            check_interval_chars=self._config.repetition_check_interval_chars,
        )

        async for event in stream:
            _raise_if_aborted(abort_flag)
            kind = str(event.get("kind", ""))
            block_index = _block_index(event)

            if kind == "text_delta":
                previous = text_by_index.get(block_index)
                if previous is None:
                    block_order.append(block_index)
                    previous = ""
                delta = str(event.get("delta", ""))
                text_by_index[block_index] = previous + delta

                repetition = repetition_detector.feed(delta)
                if repetition.detected:
                    public_events.append(
                        {
                            "type": "warning",
                            "code": "repetition_detected",
                            "message": "Repeated text was detected and the local stream was stopped.",
                        },
                    )
                    repetition_aborted = True
                    break

                public_events.append({"type": "text_delta", "delta": delta})

            elif kind == "thinking_delta":
                previous = thinking_by_index.get(block_index)
                delta = str(event.get("delta", ""))
                if previous is None:
                    block_order.append(block_index)
                    thinking_by_index[block_index] = {
                        "thinking": delta,
                        "signature": "",
                    }
                else:
                    previous["thinking"] += delta

            elif kind == "thinking_signature":
                previous = thinking_by_index.get(block_index)
                signature = str(event.get("signature", ""))
                if previous is None:
                    block_order.append(block_index)
                    thinking_by_index[block_index] = {
                        "thinking": "",
                        "signature": signature,
                    }
                else:
                    previous["signature"] = signature

            elif kind == "tool_use_start":
                tool_by_index[block_index] = _ToolInput(
                    id=str(event.get("id", "")),
                    name=str(event.get("name", "")),
                )
                block_order.append(block_index)

            elif kind == "tool_use_input_delta":
                current = tool_by_index.get(block_index)
                if current is not None:
                    current.input_json += str(event.get("partial", ""))
                    draft = _draft_candidate_from_tool(current)
                    if draft is not None:
                        preview, truncated = _preview_tail(
                            draft.content,
                            self._config.document_preview_chars,
                        )
                        state = _DocumentDraftState(
                            content_preview=preview,
                            content_length=len(draft.content),
                            truncated=truncated,
                            filename=draft.filename,
                        )
                        if document_draft_by_index.get(block_index) != state:
                            event_payload: dict[str, Any] = {
                                "type": "document_draft",
                                "id": current.id,
                                "format": draft.format,
                                "contentPreview": preview,
                                "contentLength": len(draft.content),
                                "truncated": truncated,
                            }
                            if draft.filename:
                                event_payload["filename"] = _sanitize_public_text(
                                    draft.filename,
                                )
                            public_events.append(event_payload)
                            document_draft_by_index[block_index] = state

            elif kind == "message_end":
                raw_stop = event.get("stopReason", event.get("stop_reason"))
                stop_reason = _stop_reason(raw_stop)
                raw_usage = event.get("usage")
                if isinstance(raw_usage, Mapping):
                    usage = dict(raw_usage)

            elif kind == "error":
                code = str(event.get("code", "llm_stream_error"))
                message = str(event.get("message", "stream error"))
                error = LLMStreamReaderError(code, message)
                _call_on_error(on_error, code, error)
                raise error

            _raise_if_aborted(abort_flag)

        if repetition_aborted:
            stop_reason = "end_turn"

        blocks = _assemble_blocks(
            block_order=block_order,
            text_by_index=text_by_index,
            thinking_by_index=thinking_by_index,
            tool_by_index=tool_by_index,
        )
        return LLMStreamReaderResult(
            blocks=blocks,
            public_events=public_events,
            stop_reason=stop_reason,
            usage=usage,
            provider_request=request,
        )


class _RepetitionDetector:
    def __init__(
        self,
        *,
        min_pattern_chars: int,
        threshold: int,
        check_interval_chars: int,
    ) -> None:
        self._text = ""
        self._last_check_len = 0
        self._min_pattern_chars = min_pattern_chars
        self._threshold = threshold
        self._check_interval_chars = check_interval_chars

    def feed(self, delta: str) -> _RepetitionResult:
        self._text += delta
        if len(self._text) - self._last_check_len < self._check_interval_chars:
            return _RepetitionResult(detected=False)
        self._last_check_len = len(self._text)
        return self.check()

    def check(self) -> _RepetitionResult:
        text = self._text
        if len(text) < self._min_pattern_chars * self._threshold:
            return _RepetitionResult(detected=False)

        max_pattern_len = min(len(text) // self._threshold, 500)
        candidates = list(
            range(max_pattern_len, self._min_pattern_chars + 9, -10),
        )
        candidates.extend(
            range(min(max_pattern_len, self._min_pattern_chars + 9), self._min_pattern_chars - 1, -1),
        )

        for pattern_len in candidates:
            candidate = text[-pattern_len:]
            count = 0
            search_from = 0
            while search_from <= len(text) - pattern_len:
                index = text.find(candidate, search_from)
                if index < 0:
                    break
                count += 1
                if count >= self._threshold:
                    return _RepetitionResult(
                        detected=True,
                        pattern=candidate[:80],
                        count=count,
                    )
                search_from = index + pattern_len

        return self._check_sentence_repetition(text)

    def _check_sentence_repetition(self, text: str) -> _RepetitionResult:
        sentences = [
            sentence
            for sentence in _SENTENCE_SPLIT_RE.split(text)
            if len(sentence) >= self._min_pattern_chars
        ]
        if len(sentences) < self._threshold:
            return _RepetitionResult(detected=False)

        counts: dict[str, int] = {}
        for sentence in sentences:
            normalized = " ".join(sentence.split())
            counts[normalized] = counts.get(normalized, 0) + 1

        for sentence, count in counts.items():
            if count >= self._threshold and len(sentence) >= self._min_pattern_chars:
                return _RepetitionResult(
                    detected=True,
                    pattern=sentence[:80],
                    count=count,
                )
        return _RepetitionResult(detected=False)


def _build_provider_request(
    *,
    model: str,
    messages: Sequence[Mapping[str, Any]],
    system_prompt: str,
    tool_defs: Sequence[Mapping[str, Any]],
    thinking_override: Mapping[str, Any] | None,
    trace_id: str | None,
    authoritative_model: bool,
    routing: Mapping[str, Any] | None,
    request_controlled_metadata: Mapping[str, Any] | None,
) -> dict[str, Any]:
    request: dict[str, Any] = {
        "model": model,
        "messages": [dict(message) for message in messages],
    }
    if system_prompt:
        request["system"] = system_prompt
    if tool_defs:
        request["tools"] = [dict(tool_def) for tool_def in tool_defs]
    if thinking_override is not None:
        request["thinking"] = dict(thinking_override)
    if trace_id:
        request["traceId"] = trace_id
    if authoritative_model:
        request["authoritativeModel"] = True
    if routing is not None:
        request["routing"] = dict(routing)
    if _has_request_controlled_escalation(request_controlled_metadata):
        request["requestControlledEscalationRejected"] = True
    return request


def _assemble_blocks(
    *,
    block_order: Sequence[int],
    text_by_index: Mapping[int, str],
    thinking_by_index: Mapping[int, Mapping[str, str]],
    tool_by_index: Mapping[int, _ToolInput],
) -> list[dict[str, Any]]:
    blocks: list[dict[str, Any]] = []
    seen: set[int] = set()

    for block_index in block_order:
        if block_index in seen:
            continue
        seen.add(block_index)

        if block_index in text_by_index:
            blocks.append({"type": "text", "text": text_by_index[block_index]})
            continue

        thinking = thinking_by_index.get(block_index)
        if thinking is not None:
            blocks.append(
                {
                    "type": "thinking",
                    "thinking": thinking.get("thinking", ""),
                    "signature": thinking.get("signature", ""),
                },
            )
            continue

        tool = tool_by_index.get(block_index)
        if tool is not None:
            blocks.append(
                {
                    "type": "tool_use",
                    "id": tool.id,
                    "name": tool.name,
                    "input": _parse_tool_input(tool.input_json),
                },
            )

    return blocks


def _parse_tool_input(raw: str) -> Any:
    if not raw:
        return {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return {"_malformed": True, "_raw": raw}


def _draft_candidate_from_tool(tool: _ToolInput) -> _DocumentDraftCandidate | None:
    if tool.name not in {"FileWrite", "DocumentWrite"}:
        return None

    parsed = _parsed_object(tool.input_json)
    if tool.name == "FileWrite":
        filename = _string_field(
            parsed,
            ("path", "file_path", "filepath", "filename"),
        ) or _first_json_string_field_prefix(
            tool.input_json,
            ("path", "file_path", "filepath", "filename"),
        )
        file_format = _document_format_from_path(filename)
        if file_format is None:
            return None
        content = _string_field(parsed, ("content",)) or _json_string_field_prefix(
            tool.input_json,
            "content",
        )
        if not content:
            return None
        return _DocumentDraftCandidate(
            filename=filename,
            format=file_format,
            content=content,
        )

    filename = _string_field(parsed, ("filename", "path")) or _first_json_string_field_prefix(
        tool.input_json,
        ("filename", "path"),
    )
    format_value = _string_field(parsed, ("format",)) or _json_string_field_prefix(
        tool.input_json,
        "format",
    )
    file_format = (
        format_value
        if format_value in {"md", "txt"}
        else _document_format_from_path(filename)
    )
    if file_format not in {"md", "txt"}:
        return None

    source = parsed.get("source") if parsed is not None else None
    content = _document_write_source_content(source) or _first_json_string_field_prefix(
        tool.input_json,
        ("content", "markdown", "text", "source"),
    )
    if not content:
        return None
    return _DocumentDraftCandidate(
        filename=filename,
        format=file_format,
        content=content,
    )


def _document_format_from_path(path: str | None) -> Literal["md", "txt"] | None:
    normalized = (path or "").split("?", 1)[0].split("#", 1)[0].lower()
    if normalized.endswith((".md", ".markdown")):
        return "md"
    if normalized.endswith(".txt"):
        return "txt"
    return None


def _document_write_source_content(source: Any) -> str | None:
    if isinstance(source, str):
        return source
    if isinstance(source, Mapping):
        return _string_field(source, ("content", "markdown", "text"))
    return None


def _parsed_object(raw: str) -> dict[str, Any] | None:
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        return None
    if isinstance(parsed, dict):
        return parsed
    return None


def _string_field(
    value: Mapping[str, Any] | None,
    keys: Sequence[str],
) -> str | None:
    if value is None:
        return None
    for key in keys:
        field_value = value.get(key)
        if isinstance(field_value, str):
            return field_value
    return None


def _first_json_string_field_prefix(raw: str, keys: Sequence[str]) -> str | None:
    for key in keys:
        value = _json_string_field_prefix(raw, key)
        if value is not None:
            return value
    return None


def _json_string_field_prefix(raw: str, key: str) -> str | None:
    escaped_key = re.escape(key)
    pattern = re.compile(rf'"{escaped_key}"\s*:\s*"')
    match = pattern.search(raw)
    if match is None:
        return None
    return _decode_json_string_prefix(raw, match.end())


def _decode_json_string_prefix(raw: str, start: int) -> str:
    value = ""
    escaped = False
    index = start
    while index < len(raw):
        char = raw[index]
        if escaped:
            if char == "n":
                value += "\n"
            elif char == "r":
                value += "\r"
            elif char == "t":
                value += "\t"
            elif char == "b":
                value += "\b"
            elif char == "f":
                value += "\f"
            elif char == "u":
                hex_value = raw[index + 1 : index + 5]
                if re.fullmatch(r"[0-9a-fA-F]{4}", hex_value):
                    value += chr(int(hex_value, 16))
                    index += 4
            else:
                value += char
            escaped = False
        elif char == "\\":
            escaped = True
        elif char == '"':
            break
        else:
            value += char
        index += 1
    return value


def _preview_tail(content: str, limit: int) -> tuple[str, bool]:
    sanitized = _sanitize_public_text(content)
    if limit < 1:
        return "", bool(sanitized)
    if len(sanitized) <= limit:
        return sanitized, False
    return sanitized[-limit:], True


def _sanitize_public_text(value: str) -> str:
    return "".join(
        char
        for char in value
        if char in "\n\r\t" or ord(char) >= 32
    )


def _block_index(event: Mapping[str, Any]) -> int:
    raw = event.get("blockIndex", event.get("block_index", 0))
    if isinstance(raw, int):
        return raw
    try:
        return int(str(raw))
    except ValueError:
        return 0


def _stop_reason(value: Any) -> StopReason | None:
    if value in {
        "end_turn",
        "tool_use",
        "max_tokens",
        "stop_sequence",
        "refusal",
        "pause_turn",
    }:
        return value
    return None


def _has_request_controlled_escalation(
    metadata: Mapping[str, Any] | None,
) -> bool:
    if metadata is None:
        return False
    for key in _ESCALATION_KEYS:
        value = metadata.get(key)
        if isinstance(value, str) and value.strip():
            return True
    return False


def _raise_if_aborted(flag: object | None) -> None:
    if flag is None:
        return
    aborted = False
    for attr in ("aborted", "cancelled"):
        value = getattr(flag, attr, None)
        if callable(value):
            aborted = bool(value())
        elif value is not None:
            aborted = bool(value)
        if aborted:
            break
    if not aborted:
        is_set = getattr(flag, "is_set", None)
        if callable(is_set):
            aborted = bool(is_set())
    if not aborted:
        return

    reason = getattr(flag, "reason", None)
    if isinstance(reason, BaseException):
        raise reason
    raise LLMStreamAbortError("llm_stream_aborted")


def _call_on_error(
    on_error: OnError | None,
    code: str,
    error: BaseException,
) -> None:
    if on_error is not None:
        on_error(code, error)
